mitodl/bootcamp-ecommerce
# Port of authentication app and friends

#### What are the relevant tickets?
Part of #400 
Part of #379 

#### What's this PR do?
This PR adds backend support for registration and login via email/password. It is largely a port of the relevant applications from mitxpro. At a high level, the `authentication`, `compliance`, and `user` (`profile` in this project) apps were copied over, plus any related utilities or other code that was used by those applications, hence the extremely large diff here.

Additionally, `settings.py` was updated to use the new `env` module from xpro as well, since so many new vars were brought over it was worth making sure that our configuration was more robust.

Some adjustments were made in the `authentication` app, particularly `authentication.pipeline` to support both edx and email login simultaneously, this was required because the social auth actions written for xpro were built expecting to be called from a restful api. Instead, they now check to see if the API is being invoked in a traditional browser flow or via the API, redirecting to the URLs where our react auth frontend will be when we add that.

#### How should this be manually tested?

- Test the app without making any configuration changes, everything should still work, particularly:
  - Login via edX
  - Registration via edX
  - SMApply integration
  - Making payments
- Next, test the app with `FEATURE_SOCIAL_AUTH_API=True`:
  - If you register an edx user, you should be redirected to `/create-account/extra?partial_token=TOKEN`, this page isn't functional, we just want to be sure the redirect happens.
  - For an existing edx user, set `Profile.is_complete` to `True`, and login, you should end up on the dashboard as usual.

Note: we're not going to test the compliance app in this PR, it is pretty much entirely a copy/paste so there's not much worry and it'll be disabled in all envs anyway until we configure it. It's very well covered by tests in terms of the auth flow though, so it's not a huge concern.
